cpp/daal/src/algorithms/dtrees/gbt/regression/gbt_regression_predict_dense_default_batch_impl.i

```diff
@@ -109,44 +114,120 @@ protected:
+                        size_t resIncrement)
     {
         const bool hasAnyMissing = checkForMissing(x, nTrees, nRows, nColumns);
         if (hasAnyMissing)
         {
-            predict<true, vectorBlockSize>(iTree, nTrees, nRows, nColumns, x, res);
+            predict<true, vectorBlockSize>(iTree, nTrees, nRows, nColumns, x, res, resIncrement);
+        }
+        else
+        {
+            predict<false, vectorBlockSize>(iTree, nTrees, nRows, nColumns, x, res, resIncrement);
+        }
+    }
+
+    template <bool hasUnorderedFeatures, bool hasAnyMissing>
+    inline services::Status predictContributions(size_t iTree, size_t nTrees, size_t nRowsData, const algorithmFPType * x, algorithmFPType * res,
+                                                 int condition, FeatureIndexType conditionFeature, const DimType & dim);
+
+    template <bool hasAnyMissing>
+    inline services::Status predictContributions(size_t iTree, size_t nTrees, size_t nRowsData, const algorithmFPType * x, algorithmFPType * res,
+                                                 int condition, FeatureIndexType conditionFeature, const DimType & dim)
+    {
+        if (_featHelper.hasUnorderedFeatures())
+        {
+            return predictContributions<true, hasAnyMissing>(iTree, nTrees, nRowsData, x, res, condition, conditionFeature, dim);
+        }
+        else
+        {
+            return predictContributions<false, hasAnyMissing>(iTree, nTrees, nRowsData, x, res, condition, conditionFeature, dim);
+        }
+    }
+
+    // TODO: Add vectorBlockSize templates, similar to predict
+    // template <size_t vectorBlockSize>
+    inline services::Status predictContributions(size_t iTree, size_t nTrees, size_t nRowsData, const algorithmFPType * x, algorithmFPType * res,
+                                                 int condition, FeatureIndexType conditionFeature, const DimType & dim)
+    {
+        const size_t nColumnsData = dim.nCols;
+        const bool hasAnyMissing  = checkForMissing(x, nTrees, nRowsData, nColumnsData);
+        if (hasAnyMissing)
+        {
+            return predictContributions<true>(iTree, nTrees, nRowsData, x, res, condition, conditionFeature, dim);
+        }
+        else
+        {
+            return predictContributions<false>(iTree, nTrees, nRowsData, x, res, condition, conditionFeature, dim);
+        }
+    }
+
+    template <bool hasUnorderedFeatures, bool hasAnyMissing>
+    inline services::Status predictContributionInteractions(size_t iTree, size_t nTrees, size_t nRowsData, const algorithmFPType * x,
+                                                            const algorithmFPType * nominal, algorithmFPType * res, const DimType & dim);
+
+    template <bool hasAnyMissing>
+    inline services::Status predictContributionInteractions(size_t iTree, size_t nTrees, size_t nRowsData, const algorithmFPType * x,
```

Comment:
I would need your help with that - haven't used it before. But aren't std primitives disallowed in daal? @Vika-F

Reply:
Yes, I think we cannot use std::visit here because of STL's templates that do not match with DAAL's building scheme.
I can only think of adding something like a C-style functions' table. Which is also not an elegant solution. A pseudocode:
```cpp

/// define a pointer to the predictContributions function
typedef Status (*predictContributionsFuncPtr) (size_t, size_t, size_t, const algorithmFPType *, algorithmFPType *, int, FeatureIndexType, const DimType &);

struct predictContributionsFunctor {
    /// call the corresponding variant from the functions's table
    Status call(bool hasUnorderedFeatures, bool hasAnyMissing, size_t iTree, ..., const DimType & dim) {
       
        return ptr_[int(hasUnorderedFeatures)][int(hasAnyMissing)](iTree, ..., dim);
    }
private:
    /// table that contains all the template variants of the predictContributions function
    const predictContributionsFuncPtr ptr_[2][2] = {{
            predictContributions<false, false>,
            predictContributions<false, true>
        }, {
            predictContributions<true, false>,
            predictContributions<true, true>
        }
    };
};
```
then it can be used like:
```cpp
Status st = predictContributionsFunctor().call(_featHelper.hasUnorderedFeatures(), checkForMissing(x, nTrees, nRowsData, dim.nCols), iTree, ..., dim);
```